File: src/11.binarySearch/16.33.go
package array_greedy

func Search1(nums []int, target int) int {
	for i, v := range nums {
		if v == target {
			return i
		}
	}
	return -1
}

func Search(nums []int, target int) int {

	l := 0
	r := len(nums) - 1

	for l <= r {
		mid := l + (r-l)/2
		if nums[mid] == target {
			return mid
		}

		if nums[mid] >= nums[l] { // 在左半段
			if target >= nums[l] && target < nums[mid] {
				// target 在左半段
				r = mid - 1
			} else {
				// target 在右半段
				l = mid + 1
			}

		} else { // 在右半段
			if target > nums[mid] && target <= nums[r] {
				l = mid + 1
			} else {
				r = mid - 1
			}
		}
	}

	return -1
}

File: src/11.binarySearch/16.33.ts
export function search1(nums: number[], target: number): number {
    for (let i = 0; i < nums.length; i++) {
        if (target === nums[i]) {
            return i;
        }
    }
    return -1;
}

export function search(nums: number[], target: number): number {
    let l = 0;
    let r = nums.length - 1;

    while (l <= r) {
        const mid = l + Math.floor((r - l) / 2);
        if (nums[mid] === target) {
            return mid;
        }

        if (nums[mid] >= nums[l]) {
            if (target >= nums[l] && target < nums[mid]) {
                r = mid - 1;
            } else {
                l = mid + 1;
            }
        } else {
            if (target > nums[mid] && target <= nums[r]) {
                l = mid + 1;
            } else {
                r = mid - 1;
            }
        }
    }

    return -1;
}
